fpdf/text_region.py

```diff
@@ -0,0 +1,428 @@
+import math
+from typing import NamedTuple, Sequence
+
+from .errors import FPDFException
+from .enums import Align, XPos, YPos, WrapMode
+from .line_break import MultiLineBreak
+
+# Since Python doesn't have "friend classes"...
+# pylint: disable=protected-access
+
+
+class Extents(NamedTuple):
+    left: float
+    right: float
+
+
+class TextRegionMixin:
+    """Mix-in to be added FPDF() in order to support text regions."""
+
+    def __init__(self, *args, **kwargs):
+        self.clear_text_region()
+        super().__init__(*args, **kwargs)
+
+    def register_text_region(self, region):
+        self.__current_text_region = region
+
+    def is_current_text_region(self, region):
+        return self.__current_text_region == region
+
+    def clear_text_region(self):
+        self.__current_text_region = None
+
+
+class Paragraph:
+    def __init__(
+        self,
+        region,
+        align=None,
+        line_height=None,
+        top_margin: float = 0,
+        bottom_margin: float = 0,
+        skip_leading_spaces: bool = False,
+        wrapmode: WrapMode = None,
+    ):
+        self._region = region
+        self.pdf = region.pdf
```

Comment:
What rule do you follow to decide which properties should be private, with a leading underscore, and which ones should be public?

Reply:
Roughly as random as in the rest of the code base... :wink:

Reply:
Then I would recommend to not bother with underscore-led naming, and just have public fields 🙂